File: app/airports_db.py
"""ICAO airport code → human-readable name / location.

Reads the OurAirports CSV (public domain). Data at
https://github.com/davidmegginson/ourairports-data — we parse at startup
and keep a dict in memory. Looked up by ICAO ident, which is what adsbdb
returns in `origin.icao_code` / `destination.icao_code`.

The file is baked into the Docker image at build time. If a user-provided
copy exists at /data/airports.csv it wins (same override pattern as the
aircraft DB), otherwise we fall back to the baked-in version.
"""

import csv
import logging
from pathlib import Path
from typing import Any

log = logging.getLogger("beast.airports_db")

DEFAULT_PATHS: list[Path] = [
    Path("/data/airports.csv"),
    Path(__file__).parent / "airports.csv",
]

# Types worth keeping. Heliports, closed, and seaplane bases clutter the
# database without adding much for ADS-B route display.
KEEP_TYPES = frozenset({"small_airport", "medium_airport", "large_airport"})
# Sort priority for bbox-culled responses — bigger airports win when the
# result hits the row limit so wide views still show all the majors.
TYPE_RANK = {"large_airport": 0, "medium_airport": 1, "small_airport": 2}


class AirportsDB:
    """In-memory ICAO → {name, city, country} lookup."""

    def __init__(self) -> None:
        self._db: dict[str, dict[str, Any]] = {}

    def __len__(self) -> int:
        return len(self._db)

    def __contains__(self, icao: str) -> bool:
        return icao.upper() in self._db

    def lookup(self, icao: str | None) -> dict[str, Any] | None:
        if not icao:
            return None
        return self._db.get(icao.upper())

    def load_from(self, path: Path) -> int:
        """Parse OurAirports CSV into a fresh dict and swap in atomically."""
        new_db: dict[str, dict[str, Any]] = {}
        with path.open("rt", encoding="utf-8", newline="") as fh:
            reader = csv.DictReader(fh)
            for row in reader:
                if row.get("type") not in KEEP_TYPES:
                    continue
                ident = (row.get("ident") or "").strip().upper()
                name = (row.get("name") or "").strip()
                if not ident or not name:
                    continue
                try:
                    lat = float(row.get("latitude_deg") or "")
                    lon = float(row.get("longitude_deg") or "")
                except ValueError:
                    # Entries without coordinates are useless to us (no map
                    # rendering, no bbox inclusion), so skip them entirely.
                    continue
                new_db[ident] = {
                    "name": name,
                    "city": (row.get("municipality") or "").strip(),
                    "country": (row.get("iso_country") or "").strip(),
                    "type": row["type"],
                    "lat": lat,
                    "lon": lon,
                }
        self._db = new_db
        return len(self._db)

    def bbox(
        self,
        min_lat: float,
        min_lon: float,
        max_lat: float,
        max_lon: float,
        limit: int = 2000,
    ) -> list[dict]:
        """Return airports inside the bounding box, biggest first.

        Handles antimeridian wrap when `min_lon > max_lon`. Truncates to
        `limit` rows sorted by type (large → medium → small), so a very
        wide view still includes all major airports.
        """
        wraps = min_lon > max_lon
        hits: list[dict] = []
        for icao, info in self._db.items():
            lat = info.get("lat")
            lon = info.get("lon")
            if lat is None or lon is None:
                continue
            if not (min_lat <= lat <= max_lat):
                continue
            in_lon = (lon >= min_lon or lon <= max_lon) if wraps else (min_lon <= lon <= max_lon)
            if not in_lon:
                continue
            hits.append(
                {
                    "icao": icao,
                    "name": info["name"],
                    "lat": lat,
                    "lon": lon,
                    "type": info.get("type", ""),
                }
            )
        hits.sort(key=lambda e: TYPE_RANK.get(e["type"], 9))
        return hits[:limit]

    def load_first_available(self, paths: list[Path] | None = None) -> int:
        for p in paths or DEFAULT_PATHS:
            if p.exists():
                try:
                    n = self.load_from(p)
                    log.info("loaded airports DB from %s (%d entries)", p, n)
                    return n
                except Exception as e:
                    log.warning("failed to load airports DB from %s: %s", p, e)
        log.info("no airports DB found; tooltips disabled")
        return 0

File: dotnet/src/FlightJar.Core/ReferenceData/AirportsDb.cs
using System.Collections.Frozen;
using System.Globalization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace FlightJar.Core.ReferenceData;

/// <summary>
/// ICAO airport code → human-readable name / location. Reads the
/// <a href="https://github.com/davidmegginson/ourairports-data">OurAirports</a>
/// public-domain CSV at startup. Ports <c>app/airports_db.py</c>.
/// </summary>
public sealed class AirportsDb
{
    /// <summary>Types worth keeping. Heliports, closed, seaplane bases clutter
    /// the display without adding value for ADS-B route lookups.</summary>
    private static readonly FrozenSet<string> KeepTypes = new[]
    {
        "small_airport", "medium_airport", "large_airport",
    }.ToFrozenSet();

    private static readonly IReadOnlyDictionary<string, int> TypeRank = new Dictionary<string, int>
    {
        ["large_airport"] = 0,
        ["medium_airport"] = 1,
        ["small_airport"] = 2,
    };

    private readonly ILogger _logger;
    private FrozenDictionary<string, AirportRecord> _byIcao = FrozenDictionary<string, AirportRecord>.Empty;

    public AirportsDb(ILogger<AirportsDb>? logger = null)
    {
        _logger = logger ?? NullLogger<AirportsDb>.Instance;
    }

    public int Count => _byIcao.Count;

    public AirportRecord? Lookup(string? icao)
    {
        if (string.IsNullOrEmpty(icao))
        {
            return null;
        }
        return _byIcao.TryGetValue(icao.ToUpperInvariant(), out var r) ? r : null;
    }

    public async Task<int> LoadFromAsync(string path, CancellationToken ct = default)
    {
        using var reader = File.OpenText(path);
        var fresh = new Dictionary<string, AirportRecord>(StringComparer.Ordinal);
        await foreach (var row in CsvReader.ReadDictAsync(reader, ',', ct))
        {
            if (!row.TryGetValue("type", out var type) || !KeepTypes.Contains(type))
            {
                continue;
            }
            var ident = Trim(row, "ident").ToUpperInvariant();
            var name = Trim(row, "name");
            if (ident.Length == 0 || name.Length == 0)
            {
                continue;
            }
            if (!double.TryParse(Trim(row, "latitude_deg"), NumberStyles.Float, CultureInfo.InvariantCulture, out var lat)
                || !double.TryParse(Trim(row, "longitude_deg"), NumberStyles.Float, CultureInfo.InvariantCulture, out var lon))
            {
                continue;
            }
            fresh[ident] = new AirportRecord(
                Icao: ident,
                Name: name,
                City: Trim(row, "municipality"),
                Country: Trim(row, "iso_country"),
                Type: type,
                Lat: lat,
                Lon: lon);
        }
        _byIcao = fresh.ToFrozenDictionary(StringComparer.Ordinal);
        return _byIcao.Count;
    }

    /// <summary>Return airports inside the given bounding box, biggest-first.
    /// Handles antimeridian wrap when <paramref name="minLon"/> &gt; <paramref name="maxLon"/>.</summary>
    public List<AirportRecord> Bbox(
        double minLat, double minLon, double maxLat, double maxLon, int limit = 2000)
    {
        var wraps = minLon > maxLon;
        var hits = new List<AirportRecord>();
        foreach (var info in _byIcao.Values)
        {
            if (info.Lat < minLat || info.Lat > maxLat)
            {
                continue;
            }
            var inLon = wraps ? (info.Lon >= minLon || info.Lon <= maxLon) : (info.Lon >= minLon && info.Lon <= maxLon);
            if (!inLon)
            {
                continue;
            }
            hits.Add(info);
        }
        hits.Sort((a, b) =>
        {
            var ra = TypeRank.TryGetValue(a.Type, out var x) ? x : 9;
            var rb = TypeRank.TryGetValue(b.Type, out var y) ? y : 9;
            return ra.CompareTo(rb);
        });
        if (hits.Count > limit)
        {
            hits.RemoveRange(limit, hits.Count - limit);
        }
        return hits;
    }

    public async Task<int> LoadFirstAvailableAsync(IEnumerable<string> paths, CancellationToken ct = default)
    {
        foreach (var p in paths)
        {
            if (!File.Exists(p))
            {
                continue;
            }
            try
            {
                var n = await LoadFromAsync(p, ct);
                _logger.LogInformation("loaded airports DB from {Path} ({Count} entries)", p, n);
                return n;
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                _logger.LogWarning(ex, "failed to load airports DB from {Path}", p);
            }
        }
        _logger.LogInformation("no airports DB found; tooltips disabled");
        return 0;
    }

    private static string Trim(Dictionary<string, string> row, string key)
        => row.TryGetValue(key, out var v) ? v.Trim() : "";
}

public sealed record AirportRecord(
    string Icao,
    string Name,
    string City,
    string Country,
    string Type,
    double Lat,
    double Lon);
